### DISCUSSION TITLE
How to access all the keys of a Zod Object after .refine is applied?

### DISCUSSION THREAD

**Discussion by maneetgoyal** [colinhacks/zod#435] [Q&A] [answered]
Hi all,

I am **migrating from v1 to v3 (beta)** and running into some issues. Here how my Zod object look like:

```
SomeZod = zod
  .object({
    aa: zod.string(),
    bb: zod.number(),
    cc: zod.boolean(),
    dd: AnotherZod,
  })
  .refine(
    // code
    // more code 
  );
```

- I need access to all the keys of the schema: `SomeZod --do something--> ["aa", "bb", "cc", "dd"]`. Is it possible in Zod v3?
- If I don't use `.refine`, I can use `Object.keys(SomeZod.shape)`.
- But I don't want to get rid of the `.refine` logic.
- In v1, this issue wasn't there.
- I could use `Object.keys(SomeZod._input)` but was a bit concerned about using the "private" property.

**Comment by maneetgoyal** [comment, marked as answer]
I need the array of keys to remove all unwanted keys from the object (without really having to validate it). From v3 onwards, Zod is removing keys by default when we parse, so as a workaround I am just parsing the object when I need to "trim" it.

**Reply by colinhacks** [reply]
You're right, all refinements and transforms now run inside a dedicated ZodEffects class that wraps an inner schema. Previously refinements and transformers were stored on the instance itself. I recommend splitting up the definition:


```ts
const innerObj = zod
  .object({
    aa: zod.string(),
    bb: zod.number(),
    cc: zod.boolean(),
    dd: AnotherZod,
  });
const mySchema = innerObj.refine(
  // do stuff with Object.keys(innerObj.shape)
);
```

I also just added the `innerType` getter to ZodEffects in beta2. So now you can retrieve the inner schema from a ZodEffects instance.

```ts
z.string().refine(/* refine*/).innerType; // ZodString
```

**Reply by maneetgoyal** [reply]
Thanks @colinhacks. `innerType` looks useful in the long run.
